ElectricCar.cpp

```diff
@@ -4,13 +4,48 @@
 ElectricCar::ElectricCar(ElectricEngine* engine)
     : engine_(engine)
 {
-    std::cout << __FUNCTION__ << std::endl;
+    std::cout << "* " << __FUNCTION__ << std::endl;
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() {
+    std::cout << "* " << __FUNCTION__ << std::endl;
+    delete engine_;
+}
+
+void ElectricCar::charge() { 
+    std::cout << __FUNCTION__;
+    if (chargeLevel_ < MAX_CAPACITY) {
+        chargeLevel_ = MAX_CAPACITY;
+        std::cout << " : recharged" << std::endl;
+    } else {
+        std::cerr << " : battery already full" << std::endl;
+    }
+}
+
+size_t ElectricCar::getChargeLevel() {
+    return chargeLevel_;
+}
 
+void ElectricCar::setEngine(ElectricEngine* engine) {
+    std::cout << __FUNCTION__;
+    if (!getSpeed()) {
+        std::cout << " : engine changed" << std::endl;
+        delete engine_;
+        engine_ = engine;
+        std::cout << "@ electric power: " << engine_->getPower() << std::endl;
+    } else {
+        std::cout << " : not possible unless stopped" << std::endl;
+        delete engine;
```

Comment:
here is the problem with double delete. You are deleting the engine if replacing it fails. You shouldn't delete it here.

Reply:
I test it in various ways and sometimes have 'Invalid free() / delete / delete[] / realloc()' and at the same time 'All heap blocks were freed -- no leaks are possible' - strange! Cannot figure out why there's a leak while I pass newly created object to the 'setEngine' method but the engine cannot be changed (as the car drives). Then there's a memory leak as the object was created and did not assigned to a pointer. The only way seems to pass the newly created engine as the car is stopped to avoid leaks. But there should be another way.

Reply:
No more problems nor memory leaks after using shared_ptr. Finally!